lib/puppet/type/file/checksum_value.rb

```diff
@@ -0,0 +1,74 @@
+require 'puppet/util/checksums'
+
+module Puppet
+  Puppet::Type.type(:file).newproperty(:checksum_value) do
+    include Puppet::Util::Checksums
+    include Puppet::Util::Diff
+
+    desc "The checksum of the source contents. Only md5 and sha256 are supported when
+      specifying this parameter."
+
+    def insync?(is)
+      # If checksum_value is specified, only manage source.
+      # Content manages its own syncing... FOR NOW, duh duh duh
+      if resource.parameter(:source).nil?
+        return true
+      end
+
+      if resource.should_be_file?
+        return false if is == :absent
+      else
+        return true
+      end
+
+      return true if ! @resource.replace?
+
+      result = super
+
+      if ! result and Puppet[:show_diff] and resource.show_diff?
+        resource.write_temporarily do |path|
+          send @resource[:loglevel], "\n" + diff(@resource[:path], path)
+        end
+      end
+      result
+    end
+
+    def property_matches?(current, desired)
+      time_types = [:mtime, :ctime]
+      checksum_type = resource.parameter(:checksum).value
+
+      # The inherited equality is always accepted, so use it if valid.
+      basic = super(current, desired)
+      return basic if basic || !time_types.include?(checksum_type)
+      return false unless current && desired
+      begin
+        DateTime.parse(current) >= DateTime.parse(desired)
+      rescue => detail
+        self.fail Puppet::Error, "Resource with checksum_type #{checksum_type} didn't contain a date in #{current} or #{desired}", detail.backtrace
+      end
+    end
+
+    def sync
+      # insync? only returns false if it expects to manage the file content,
+      # so instruct the resource to write its contents.
+      return_event = @resource.stat ? :file_changed : :file_created
+      @resource.write
+      return_event
+    end
+
+    def retrieve
+      # Intentionally mirrors content#retrieve. Eventually content will be a
+      # parameter, and content syncing will be managed by checksum_value.
```

Comment:
That's a little surprising to me. I've been thinking (some day) `content` would not have any `actual_content` or magic behavior around munging the value to the checksum. Instead the munged `content` value would just be the desired value specified in the manifest, and we would overwrite the `insync?` and `property_matches?` logic to compare checksums (but without relying on the existence of the `checksum_value` property. That way `content` and `source`/`checksum_value` are completely independent, and file writing is triggered either by calling `Content#sync` or `ChecksumValue#sync`.

But if I understand correctly, you're thinking of creating a single `checksum_value` property whose value is obtained either from `content` or `source`. So all of the `insync?` logic would be in one place, but the "get the desired checksum value" could be delegated to `content` and `source`?

That seems appealing though it's unintuitive to think of `content` as a parameter. But I suppose that's what `actual_content` was doing, acting as a parameter within the property.